operator/pkg/ciliumendpointbatch/endpointbatch.go

```diff
@@ -0,0 +1,307 @@
+	), "cilium_endpoint_batch")
+
+	manager := newCebManagerFcfs(rlQueue, maxCepsInCeb)
+	cebStore := ciliumEndpointBatchInit(client.CiliumV2alpha1(), wait.NeverStop)
+
+	// List all existing CEBs from the api-server and cache it locally.
+	// This sync should happen before starting CEP watcher, because CEP watcher
+	// emits the existing CEPs as newly added CEPs. If we don't have local sync
+	// cebManager would assume those are new CEPs and may create new CEBs for those CEPs.
+	// This situation ends up having duplicate CEPs in different CEBs. Hence, we need
+	// to sync existing CEBs before starting a CEP watcher.
+	syncCebsInLocalCache(cebStore, manager)
+	cebC := &CiliumEndpointBatchController{
+		clientV2:                            client.CiliumV2(),
+		clientV2a1:                          client.CiliumV2alpha1(),
+		reconciler:                          newReconciler(client.CiliumV2alpha1(), manager),
+		Manager:                             manager,
+		queue:                               rlQueue,
+		ciliumEndpointBatchStore:            cebStore,
+		ciliumEndpointBatchUpsertSyncPeriod: cebSyncPeriod,
+		ciliumEndpointBatchDeleteSyncPeriod: cebDeleteSyncPeriod,
+	}
+	return cebC
+}
+
+func (c *CiliumEndpointBatchController) Run(ces cache.Indexer, stopCh chan struct{}) {
+
+	log.Info("Bootstrap ceb controller")
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	// Cache CiliumEndpointStore Interface locally
+	c.ciliumEndpointStore = ces
+
+	// On operator warm boot, remove stale cep entries present in CEB
+	c.ciliumRemoveStaleCepEntries()
+
+	// Reconcile current state with desired state at every
+	// ciliumEndpointBatchUpsertSyncPeriod
+	log.WithFields(logrus.Fields{
+		"Upsert syncperiod": c.ciliumEndpointBatchUpsertSyncPeriod,
+		"Delete syncperiod": c.ciliumEndpointBatchDeleteSyncPeriod,
+	}).Debug("Starting CEB controller reconciler.")
+
+	go wait.Until(c.worker, c.ciliumEndpointBatchUpsertSyncPeriod, stopCh)
+
+	go func() {
+		defer utilruntime.HandleCrash()
+	}()
+
+	<-stopCh
+
+	return
+}
+
+// Upon warm boot[restart], Iterate over all CEPs which we got from the api-server
+// and compare it with CEPs packed inside CEB.
+// If there are any stale CEPs present in CEBs, remove them from their CEB.
+func (c *CiliumEndpointBatchController) ciliumRemoveStaleCepEntries() {
+	log.Info("Remove stale CEP entries in CEB")
+
+	// Get all ceps from local datastore
+	staleCeps := c.Manager.getAllCeps()
+
+	// Remove stale CEP entries present in CEB
+	for cepName, cCep := range staleCeps {
+		if _, exists, err := c.ciliumEndpointStore.GetByKey(getCepNameFromCCEP(cCep)); err == nil && exists || err != nil {
+			continue
+		}
+		log.WithFields(logrus.Fields{
+			"cep-name": cepName,
+		}).Debug("Removing stale CEP entry.")
+		c.Manager.RemoveCepFromCache(cCep)
+	}
+	return
+}
+
+// syncs all cebs from cebStore in manager cache.
+func syncCebsInLocalCache(cebStore cache.Store, manager cebManager) {
```

Comment:
Not really, CEB controller isn't yet running. This happens before CEB controller running.

Reply:
`CEB watcher` is consumer of `CEB` updates, though we start watching it. There will be no updates unless and until `CEB controller` running.

Reply:
When you start the watcher, the store is subject to watch event and updates from apiserver, we should not rely on the assumption that we are the only writer hence no update would come. What happens if one operator pushes updates then goes down and another one comes up? The new leader might get the update even if it does not create one.

Reply:
>>What happens if one operator pushes updates then goes down and another one comes up? The new leader might get the update even if it does not create one.
`ciliumEndpointBatchInit` will wait until local cache syncs with api-server, this guarantees we get latest and greatest `CEBs` in cache.
Below snippet ensures that we synced with api-server.
`cache.WaitForCacheSync(stopCh, cebController.HasSynced)`

Reply:
I still don't think we should rely on implicit assumption here, @varunmar @aanm  WDYT?

Reply:
@aanm I'm a bit confused, how's leader election matters in this case? Even if you just have one instance, wouldn't the store subject to change once you start watcher?

Reply:
@Weil0ng The example that you have wrote here [1] sounded like you were saying that the 2 operators would concurrently write into CEBs. Assuming that we are only discussion for 1 operator, I don't think there's an issue because AFAIK this function is not pushing anything to kube-apiserver. That only happens when we receive CEP events and that watcher is not enabled at this point.

[1]
> When you start the watcher, the store is subject to watch event and updates from apiserver, we should not rely on the assumption that we are the only writer hence no update would come. What happens if one operator pushes updates then goes down and another one comes up? The new leader might get the update even if it does not create one.

Reply:
I see, yea even with leader election, you could have a pending update from a previous leader. But as @krishgobinath said, we will wait until the cache of the new leader is synced, so I agree we won't have an update during init in this case. I personally still think lock-protecting access to the watch store provides stronger guarantee, but I'm okay with current code.